src/frontend/apps/markdown/MarkdownEditor/index.tsx

```diff
@@ -0,0 +1,334 @@
+    {},
+  );
+
+  const [contentChanged, setContentChanged] = React.useState(false);
+
+  // note: we don't want to fetch the markdown document regularly to prevent
+  // any editor update while the user has not saved her document.
+  const markdownDocumentId = MarkdownAppData.markdownDocument!.id;
+  const queryClient = useQueryClient();
+  const { data: markdownDocument } = useMarkdownDocument(markdownDocumentId, {
+    refetchInterval: false,
+    refetchOnWindowFocus: false,
+  });
+  const { mutate: mutateDocument } =
+    useUpdateMarkdownDocument(markdownDocumentId);
+
+  const { mutate: saveDocumentTranslations, isLoading: isSaving } =
+    useSaveTranslations(markdownDocumentId, {
+      onSuccess: () => {
+        setContentChanged(false);
+        toast.success(intl.formatMessage(messages.saveSuccess));
+      },
+      onError: () => {
+        toast.error(intl.formatMessage(messages.saveFailure));
+      },
+    });
+
+  // Allow data reloading from backend on language change
+  useEffect(() => {
+    queryClient.invalidateQueries(['markdown-documents', markdownDocumentId]);
+  }, [language]);
+
+  // Initialization hook
+  useEffect(() => {
+    if (!markdownDocument) return;
+    const translatedTitle = getMarkdownDocumentTranslatedContent(
+      markdownDocument,
+      'title',
+      language,
+    );
+    setLocalTitle(translatedTitle);
+    setInitialTitle(translatedTitle);
+
+    const translatedContent = getMarkdownDocumentTranslatedContent(
+      markdownDocument,
+      'content',
+      language,
+    );
+    setInitialMarkdownContent(translatedContent);
+    setLocalMarkdownContent(translatedContent);
+    updateEditorContent(translatedContent);
+
+    const translatedRenderedContent = getMarkdownDocumentTranslatedContent(
+      markdownDocument,
+      'rendered_content',
+      language,
+    );
+    setLocalRenderedContent(translatedRenderedContent);
+    setInitialRenderedContent(translatedRenderedContent);
+
+    setLocalIsDraft(markdownDocument.is_draft);
+  }, [markdownDocument, language]);
+
+  // Only change detection here
+  useEffect(() => {
+    if (
+      localTitle !== initialTitle ||
+      localMarkdownContent !== initialMarkdownContent ||
+      localRenderedContent !== initialRenderedContent
+    ) {
+      setContentChanged(true);
+    } else {
+      setContentChanged(false);
+    }
+  }, [
+    localTitle,
+    localMarkdownContent,
+    localRenderedContent,
+    initialTitle,
+    initialMarkdownContent,
+    initialRenderedContent,
+  ]);
+
+  // Detect change to alert user before leaving
+  useEffect(() => {
+    const beforeUnload = (event: BeforeUnloadEvent) => {
+      if (contentChanged) {
+        event.preventDefault();
+        event.returnValue = '';
+      }
+    };
+    window.addEventListener('beforeunload', beforeUnload);
+    return () => window.removeEventListener('beforeunload', beforeUnload);
+  }, [contentChanged]); // We highly rely on any change performed
```

Comment:
I don't know if using `contentChanged` as a dependency is a good idea here. You will attach a new `eventListener` every time this value changes. This is not really what we want.
I suggest you to do something like this:

```
const beforeUnload = (event: BeforeUnloadEvent) => {
      if (contentChanged) {
        event.preventDefault();
        event.returnValue = '';
      }
    };
useEffect(() => {
    window.addEventListener('beforeunload', beforeUnload);
    return () => window.removeEventListener('beforeunload', beforeUnload);
  }, []);
```

You add only once the event lister and the callback live outside the `useEffect`.

Reply:
Yes, already tried to do like this, but it does not work as `contentChanged` is a state and is "async changed" and during my tests it was always falsy even if the component renders a `contentChanged` truthy display...
This is why, it seems to be required to `addEventListener`/`removeEventListener` at any change... I'm not happy with this but I didn't find a way to properly have the "synced" value of `contentChanged` using the very same code as in your snippet.

Reply:
Finally solved, using a `Ref` instead of an unnecessary `State` for `contentChanged`, many thanks to @RVassili  : 
```
  // Detect change to alert user before leaving
  useEffect(() => {
    const beforeUnload = (event: BeforeUnloadEvent) => {
      if (contentChanged.current) {
        event.preventDefault();
        event.returnValue = '';
      }
    };
    window.addEventListener('beforeunload', beforeUnload);
    return () => window.removeEventListener('beforeunload', beforeUnload);
  }, []); // We highly rely on any change performed
```